File: Sources/DiskMapCore/ChartLayout.swift
import Foundation

/// One drawable slice of a chart. `nodeID` is nil for the collapsed
/// remainder. That remainder is never drillable.
public struct ChartSlice: Sendable, Equatable, Identifiable {
    public let id: String
    public let nodeID: Int32?
    public let size: Int64
    public let label: String
    public let drillable: Bool
    public let children: [ChartSlice]

    public init(id: String, nodeID: Int32?, size: Int64, label: String, drillable: Bool, children: [ChartSlice]) {
        self.id = id
        self.nodeID = nodeID
        self.size = size
        self.label = label
        self.drillable = drillable
        self.children = children
    }
}

/// One level of `currentNode`'s children, plus at most one more level
/// for the children that are large enough to draw. Anything smaller
/// than 0.5% of its parent collapses into a single Other slice so a
/// home scan cannot draw every descendant.
public enum ChartLayout {
    public static let otherFraction = 0.005

    public static func slices(
        of node: Int32,
        in tree: FileTree,
        totals: [Int64],
        otherFraction fraction: Double = otherFraction
    ) -> [ChartSlice] {
        guard node >= 0, node < tree.count, totals.count == tree.count else { return [] }
        return collapse(
            tree.children(of: node, totals: totals),
            parentSize: totals[Int(node)],
            tree: tree,
            totals: totals,
            includeChildren: true,
            idPrefix: "\(node)",
            otherFraction: fraction
        )
    }

    private static func collapse(
        _ items: [(id: Int32, size: Int64)],
        parentSize: Int64,
        tree: FileTree,
        totals: [Int64],
        includeChildren: Bool,
        idPrefix: String,
        otherFraction: Double
    ) -> [ChartSlice] {
        let threshold = Double(parentSize) * otherFraction
        var visible: [(id: Int32, size: Int64)] = []
        var otherSize: Int64 = 0
        var otherCount = 0
        for item in items where item.size > 0 {
            if Double(item.size) < threshold {
                otherSize += item.size
                otherCount += 1
            } else {
                visible.append(item)
            }
        }
        visible.sort { $0.size > $1.size }
        var slices: [ChartSlice] = visible.map { item in
            let nested: [ChartSlice]
            if includeChildren, tree.isDirectory[Int(item.id)] {
                nested = collapse(
                    tree.children(of: item.id, totals: totals),
                    parentSize: item.size,
                    tree: tree,
                    totals: totals,
                    includeChildren: false,
                    idPrefix: "\(idPrefix).\(item.id)",
                    otherFraction: otherFraction
                )
            } else {
                nested = []
            }
            return ChartSlice(
                id: "\(idPrefix).\(item.id)",
                nodeID: item.id,
                size: item.size,
                label: tree.name(of: item.id),
                drillable: tree.isDirectory[Int(item.id)],
                children: nested
            )
        }
        if otherCount > 0 {
            slices.append(ChartSlice(
                id: "\(idPrefix).other",
                nodeID: nil,
                size: otherSize,
                label: "Other (\(otherCount))",
                drillable: false,
                children: []
            ))
        }
        return slices
    }
}

File: windows/src/DiskMap.Core/Layout/ChartLayout.cs
namespace DiskMap.Core;

/// <summary>
/// One drawable slice of a chart. <see cref="NodeID"/> is null for the
/// collapsed remainder. That remainder is never drillable.
/// </summary>
public sealed record ChartSlice(
    string Id,
    int? NodeID,
    long Size,
    string Label,
    bool Drillable,
    List<ChartSlice> Children);

/// <summary>
/// One level of <paramref name="node"/>'s children, plus at most one more
/// level for the children that are large enough to draw. Anything smaller
/// than 0.5% of its parent collapses into a single Other slice so a home
/// scan cannot draw every descendant.
/// </summary>
public static class ChartLayout
{
    public const double OtherFraction = 0.005;

    public static List<ChartSlice> SlicesOf(int node, FileTree tree, long[] totals)
    {
        if (node < 0 || node >= tree.Count || totals.Length != tree.Count)
            return [];
        return Collapse(tree.ChildrenOf(node, totals), totals[node], tree, totals, includeChildren: true, idPrefix: node.ToString());
    }

    private static List<ChartSlice> Collapse(
        List<(int Id, long Size)> items,
        long parentSize,
        FileTree tree,
        long[] totals,
        bool includeChildren,
        string idPrefix)
    {
        double threshold = parentSize * OtherFraction;
        var visible = new List<(int Id, long Size)>();
        long otherSize = 0;
        int otherCount = 0;
        foreach (var item in items)
        {
            if (item.Size <= 0) continue;
            if (item.Size < threshold) { otherSize += item.Size; otherCount++; }
            else visible.Add(item);
        }
        visible.Sort((a, b) => b.Size.CompareTo(a.Size));

        var slices = visible.Select(item =>
        {
            List<ChartSlice> nested = [];
            if (includeChildren && tree.IsDirectory[item.Id])
            {
                nested = Collapse(
                    tree.ChildrenOf(item.Id, totals), item.Size, tree, totals,
                    includeChildren: false, idPrefix: $"{idPrefix}.{item.Id}");
            }
            return new ChartSlice(
                Id: $"{idPrefix}.{item.Id}",
                NodeID: item.Id,
                Size: item.Size,
                Label: tree.NameOf(item.Id),
                Drillable: tree.IsDirectory[item.Id],
                Children: nested);
        }).ToList();

        if (otherCount > 0)
        {
            slices.Add(new ChartSlice(
                Id: $"{idPrefix}.other",
                NodeID: null,
                Size: otherSize,
                Label: $"Other ({otherCount})",
                Drillable: false,
                Children: []));
        }
        return slices;
    }
}
